### ISSUE TITLE
Make bash alias importing optional?

### ISSUE THREAD

**Issue by mbollmann** [scopatz/xonsh#102]
I really like that bash aliases can be imported automatically, but unfortunately it broke xonsh for me...

The reason is that I have an alias for `cd` which refers to a bash function defined in my .bashrc file. In xonsh, this function -- of course -- doesn't exist. That means I can't use the `cd` command anymore!

My current workaround is `from xonsh.aliases import cd; aliases['cd'] = cd` in .xonshrc, but I wondered if there shouldn't also be an option to disable the automatic importing from bash. (I guess `cd` is not the only broken alias for me, so I'd probably rather define my aliases by hand.)

Does a command-line switch `--no-bash-aliases` (or similar) sound reasonable? I could make a PR for it then.

**Comment by wrywerytwreywery** [comment]
I agree here; I'd rather just write my aliases in xonsh anyway (switching cold turkey from bash).

A command-line switch might work, but would that work with chsh?  That is, would it work to set the shell to be `/usr/local/bin/xonsh --no-bash-aliases`?

One option would be not to automatically import ever, but to offer a function that, when called, imports the bash aliases.  Then importing could be as easy as calling, say, `import_bash_aliases()` in the xonshrc.

Another, depending on when the import from bash is happening, might be to toggle that importing by setting a variable in the xonshrc (I haven't looked at the code for this yet, so I'm not sure whether that would work).

**Comment by scopatz** [comment]
As a temporary fix,  you should be able to delete the alias in the xonshrc. Also broken aliases should not cause crashes.

**Comment by mbollmann** [comment]
The broken alias doesn't crash; it merely fails because the function it's replaced with doesn't exist. Deleting the alias doesn't work.

```bash
bollmann@underberg ~ $ cd data
xonsh: subprocess mode: command not found: stack_cd

bollmann@underberg ~ $ del aliases['cd']
bollmann@underberg ~ $ cd data
xonsh: subprocess mode: command not found: cd
Did you mean one of the following?
    d: Alias
    dc: Command (/usr/bin/dc)
    dd: Command (/bin/dd)
    cc: Command (/usr/bin/cc)
    e: Alias
```

**Comment by scopatz** [comment]
oh right. For cd you need to do 

```python
from xonsh.aliases import  cd
aliases['cd'] = cd
```

**Comment by mbollmann** [comment]
@scopatz, yes, that's what I said in the first post ;-)

@wrywerytwreywery, I couldn't find anything on using command-line switches with `chsh`, but bash uses a symbolic link `rbash` to do the equivalent of `bash --restricted`, so I suppose it doesn't work.

In that case, either not importing bash aliases by default, or providing a variable that can be modified to turn off alias importing (bash has `$SHELLOPTS`, could we use something similar?) seems sensible to me.

**Comment by scopatz** [comment]
@mbollmann haha whoops.  I was still not really awake when I wrote this. :).  

First, I am all in favor of making this an optional, though I think it should be on by default.  We might also want to add a 'safe' mode (also on by default) that prevents and warns when bash aliases would overwrite built-ins (cd, exit, etc).  Lastly, I think the optional behaviour that should be controlled via an environment variable, preferably set in the xonshrc.  Those are my thoughts on it, at least.

**Comment by scopatz** [comment]
Hey @mbollmann - Do you wanna take a stab at this?  And if you do, do you think you'll have a chance to get a PR in before a release tomorrow night? Just asking :)

**Comment by mbollmann** [comment]
I still have this on my radar and can take a look at it tomorrow, if it's not too difficult I might get it in, but no promises since I haven't started yet... :)

**Comment by scopatz** [comment]
Haha ok! Thanks!

**Comment by mbollmann** [comment]
Hmm, looks easy, but might not be.

When aliases can potentially override each other, I believe the desired order is:

1) `DEFAULT_ALIASES`
2) Aliases imported from bash
3) Aliases defined in .xonshrc

...where later ones can override previous ones. But if we want to control the alias importing from a variable set in .xonshrc, we'd need to do the importing after processing .xonshrc, thus switching 2) and 3) -- we'd override aliases from .xonshrc with bash-imported ones, which seems undesirable to me.

Right now I don't have any bright ideas how to get around this ...

**Comment by scopatz** [comment]
Hmm you are totally right @mbollmann. This is trickier than I thought.

**Comment by SmartViking** [comment]
One idea is to make a `.weak_update` method on `Aliases`, that doesn't touch keys that are already present and update bash-imports with that. That implies that `DEFAULT_ALIASES` takes precedence over bash-aliases, not sure if that's desirable.

**Comment by scopatz** [comment]
Well, the problem to me is that bash aliases can be functions, which render them totally invalid.  Unfortunately, there is no good way I know of (other than having a bash syntax tree) to look an arbitrary string of bash code and tell whether a bash function is being used.

**Comment by scopatz** [comment]
Here is what I think we should do.  There should be an environment variable that controls this behaviour, maybe `$LOAD_BASH_ALIASES`.  We should not change the order of alias resolution that @mbollmann described.  So the challenge becomes setting the env variable before alias resolution, which makes xonshrc impossible. From here, there are three options I see, one or all of them should be supported.

1. have folks `export LOAD_BASH_ALIASES=''` in their `bashrc`.  We read in bashrc before aliases, so this is actually another run control file for us.
2. Add `--load-bash-aliases` and `--no-load-bash-aliases` to the xonsh argument parser.  These would simply set the environment variable.
3. have a xonsh environment config file file, possibly in JSON, that can be used to set environment variables as the last step in the `xonsh.environ.default_env()` function.

I don't think that this is the last time such a situation will arise, so it would be nice to have good strategy.  (1) seems a little hacky but would work now.  (2) seems like it could be annoying to use all the time. (3) seems like it is adding yet another another config file, but is otherwise clean.

What do you all think?

**Comment by wrywerytwreywery** [comment]
> 2. Add `--load-bash-aliases` and `--no-load-bash-aliases` to the xonsh argument parser. These would simply set the environment variable.

@mbollmann and I discussed this a little bit earlier in the thread.  The downside here would be that I'm not sure it would work with xonsh as the default shell (via `chsh`); that is, I'm not sure one could set the default shell to be, e.g., `/usr/local/bin/xonsh --no-load-bash-aliases`.  One workaround would be to create a second script with a different name that starts up with the `--no-load-bash-aliases` option, but that seems like a big kludge.

Would it not be possible to preserve the resolution order from above, even if the aliases aren't loaded in that order?  That is, could we:
* load the default aliases into `aliases`
* run the xonshrc, loading the results into a different dict
* if the `IMPORT_BASH_ALIASES` environment variable is True, load the bash aliases
* update `aliases` with the bash-loaded ones, if necessary
* update `aliases` with the xonshrc results

Would that work, or am I missing something?

**Comment by scopatz** [comment]
This would definitely be possible, it just wouldn't apply to any situation other than the aliases. Also, it shifts responsibility a bit.  Right now the Shell loads the xonshrc, not the `load_builtins()`. Perhaps this should be refactored anyway.  

In any event, I guess my main point here was that I would rather add another stage to the loading procedure via command line arguments and/or other config files than to perform lazy application of loading.  This is mostly because I foresee a bunch of potential situations where adjustments to how xonsh starts up are possible.  For example, we could also import aliases from zsh.  Or maybe we don't want to load bashrc at all.  Allowing for a first pass, static environment config file would allow for modification of all xonsh loading mechanism, not just bash aliases. My two cents :)

**Comment by mbollmann** [comment]
Logically, I think this behaviour is best implemented as a command-line switch, and I don't really like the added complexity of a second configuration file. People will find it more confusing at first what goes into which file. I wouldn't find a proxy script that just calls `xonsh --no-load-bash-aliases` too kludgy, but of course this also isn't a solution if such cases pop up more frequently.

Can we do both maybe? Command-line switches that can optionally be activated by a second configuration file as well?

**Comment by scopatz** [comment]
> Can we do both maybe? Command-line switches that can optionally be activated by a second configuration file as well?

Yep, I am happy with both.

**Comment by scopatz** [comment]
And they can be implemented separately :)

**Comment by scopatz** [comment]
This has been true for a while now through the static config file.  Please reopen if needed.

**Comment by ahundt** [comment]
What's the correct way to fix this? I'm running into this problem in #2486
